sevntu-checks/src/main/java/com/github/sevntu/checkstyle/checks/coding/Jsr305AnnotationsCheck.java

```diff
@@ -0,0 +1,1086 @@
+         * @param search
+         *        the annotations to look for
+         */
+        protected void checkContainsAll(final String msg, final NullnessAnnotation... search) {
+            if (!violationFound && containsAll(search)) {
+                violation(msg);
+            }
+        }
+
+        /**
+         * Emits a violation if both this and the parent class have redundant nullness annotations.
+         * @param msg
+         *        the violation message to emit
+         * @param search
+         *        the annotations to look for
+         */
+        protected void checkRedundancyDueToClassLevelAnnotation(final String msg,
+                final NullnessAnnotation... search) {
+            if (!violationFound) {
+                for (final NullnessAnnotation nullnessAnnotation : search) {
+                    final boolean thisIsAnnotated = this.annotations.contains(nullnessAnnotation);
+                    final boolean parentIsAnnotated =
+                            getParentMethodOrClassAnnotation(nullnessAnnotation) != null;
+                    if (thisIsAnnotated && parentIsAnnotated) {
+                        violation(msg);
+                        break;
+                    }
+                }
+            }
+        }
+
+        /**
+         * Check whether all the given annotations are present.
+         * @param search
+         *        the annotations to look for
+         * @return true if yes
+         */
+        private boolean containsAll(final NullnessAnnotation... search) {
+            Boolean result = null;
+            if (this.annotations.isEmpty()) {
+                // an empty list of annotations can never contain all
+                result = false;
+            }
+            else {
+                for (final NullnessAnnotation obj : search) {
+                    if (!this.annotations.contains(obj)) {
+                        result = false;
+                        break;
+                    }
+                }
+            }
+            if (result == null) {
+                result = true;
+            }
+            return result;
+        }
+
+        /**
+         * Make sure that none of the given annotations are present.
+         * @param msg
+         *        the violation message to emit if one of the given annotations was found
+         * @param search
+         *        the annotations to look for
+         */
+        protected void checkContainsNone(final String msg,
+                final NullnessAnnotation... search) {
+            if (!violationFound && !containsAny(search)) {
+                violation(msg);
+            }
+        }
+
+        /**
+         * Handle a violation (log and set 'violationFound' to 'true').
+         *
+         * @param msg
+         *        the violation message
+         */
+        private void violation(final String msg) {
+            log(ast, msg);
+            violationFound = true;
```

Comment:
There are cases in which the check (correctly) finds two different violations in one place, e.g. if a parameter has a nullness annotation not allowed in that context, then this would technically trigger two warnings: one for illegal annotation, and one for a missing legal annotation. In that case only the "illegal annotation" violation is flagged, because the actual problem is not a missing, but the wrong annotation.

I understand that as it is the code looks a bit misleading. I have added some additional Javadoc to the definition of `violationFound`. I hope that's OK?